Remove whitespace from numeric fields when processing data from the OTP Server Directory

I added a new server to the [OTP Server Directory](https://docs.google.com/spreadsheet/ccc?key=0AgWy8ujaGosCdDhxTC04cUZNeHo0eGFBQTBpU2dxN0E&usp=drive_web&authkey=CK-H__IP&authkey=CK-H__IP#gid=0) today, and accidentally left whitespace in between two numbers (after the comma `43.54, 10.8418`) in the "Default Center" field.

When I refreshed the servers list from the app, I got the following exception:

```
java.lang.RuntimeException: An error occured while executing doInBackground()
            at android.os.AsyncTask$3.done(AsyncTask.java:299)
            at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:352)
            at java.util.concurrent.FutureTask.setException(FutureTask.java:219)
            at java.util.concurrent.FutureTask.run(FutureTask.java:239)
            at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:230)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1080)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:573)
            at java.lang.Thread.run(Thread.java:841)
     Caused by: java.lang.NumberFormatException: Invalid double: " 10.8418"
            at java.lang.StringToReal.invalidReal(StringToReal.java:63)
            at java.lang.StringToReal.parseDouble(StringToReal.java:269)
            at java.lang.Double.parseDouble(Double.java:295)
            at edu.usf.cutr.opentripplanner.android.model.Server.setCenter(Server.java:164)
            at edu.usf.cutr.opentripplanner.android.model.Server.<init>(Server.java:94)
            at edu.usf.cutr.opentripplanner.android.tasks.ServerSelector.downloadServerList(ServerSelector.java:220)
            at edu.usf.cutr.opentripplanner.android.tasks.ServerSelector.doInBackground(ServerSelector.java:138)
            at edu.usf.cutr.opentripplanner.android.tasks.ServerSelector.doInBackground(ServerSelector.java:1)
            at android.os.AsyncTask$2.call(AsyncTask.java:287)
            at java.util.concurrent.FutureTask.run(FutureTask.java:234)
            at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:230)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1080)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:573)
            at java.lang.Thread.run(Thread.java:841)
```

It took Google a minute or so to make the new spreadsheet contents without the space available, and during this time I couldn't use the app.

We should remove whitespace automatically from any numeric fields retrieved from the OTP Server Directory to avoid issues like this, since maintainers of servers are likely to make this mistake when editing data in the spreadsheet.

Current numeric fields are "Bounds", "Default Center", and "Default Zoom Level".  We should also look at removing leading and trailing whitespace from the other fields.

--- a/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/exceptions/ServerListParsingException.java
+++ b/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/exceptions/ServerListParsingException.java
@@ -1,0 +1,11 @@
+package edu.usf.cutr.opentripplanner.android.exceptions;
+
+/**
+ * Created by foucelhas on 2/10/14.
+ */
+public class ServerListParsingException extends IllegalArgumentException {
+
+    public ServerListParsingException(String message) {
+        super("Problems parsing server fields: " + message);
+    }
+}

--- a/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/model/Server.java
+++ b/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/model/Server.java
@@ -20,6 +20,7 @@ import android.content.Context;
 import android.location.Location;
 
 import edu.usf.cutr.opentripplanner.android.R;
+import edu.usf.cutr.opentripplanner.android.exceptions.ServerListParsingException;
 
 /**
  * Modified by Khoa Tran
@@ -77,7 +78,7 @@ public class Server {
         super();
     }
 
-    public Server(Server s) {
+    public Server(Server s) throws ServerListParsingException {
         super();
         setId(s.getId());
         setDate(s.getDate());
@@ -92,7 +93,8 @@ public class Server {
     }
 
     public Server(Long d, String region, String baseURL, String bounds,
-            String language, String contactName, String contactEmail, String center, String zoom) {
+            String language, String contactName, String contactEmail, String center, String zoom)
+            throws ServerListParsingException {
         super();
         setDate(d);
         setRegion(region);
@@ -106,7 +108,8 @@ public class Server {
     }
 
     public Server(String region, String baseURL, String bounds,
-            String language, String contactName, String contactEmail, String center, String zoom) {
+            String language, String contactName, String contactEmail, String center, String zoom)
+            throws ServerListParsingException {
         super();
         setRegion(region);
         setBaseURL(baseURL);
@@ -156,39 +159,60 @@ public class Server {
         return zoom;
     }
 
-    public void setZoom(String zoom) {
-        zoomSet = true;
-        this.zoom = zoom;
-
-        setInitialZoom(Float.parseFloat(zoom));
+    public void setZoom(String zoom) throws ServerListParsingException {
+        try {
+            Float zoomFloat = Float.parseFloat(zoom);
+            setInitialZoom(zoomFloat);
+            zoomSet = true;
+            this.zoom = zoom;
+        } catch (NumberFormatException e) {
+            throw new ServerListParsingException("Incorrect zoom value: " + e.toString());
+        }
     }
 
     public String getCenter() {
         return center;
     }
 
-    public void setCenter(String center) {
-        centerSet = true;
-        this.center = center;
+    public void setCenter(String center) throws ServerListParsingException {
         String[] tokens = center.split(",");
-
-        setCenterLatitude(Double.parseDouble(tokens[0]));
-        setCenterLongitude(Double.parseDouble(tokens[1]));
+        if (tokens.length == 2) {
+            centerSet = true;
+            this.center = center;
+            try {
+                setCenterLatitude(Double.parseDouble(tokens[0]));
+                setCenterLongitude(Double.parseDouble(tokens[1]));
+            } catch (NumberFormatException e) {
+                throw new ServerListParsingException(
+                        "Incorrect center coordinates: " + e.toString());
+            }
+        } else {
+            throw new ServerListParsingException("Incorrect center coordinates");
+        }
     }
 
     public String getBounds() {
         return bounds;
     }
 
-    public void setBounds(String bounds) {
-        boundsSet = true;
-        this.bounds = bounds;
+    public void setBounds(String bounds) throws ServerListParsingException {
         String[] tokens = bounds.split(",");
 
-        setLowerLeftLatitude(Double.parseDouble(tokens[0]));
-        setLowerLeftLongitude(Double.parseDouble(tokens[1]));
-        setUpperRightLatitude(Double.parseDouble(tokens[2]));
-        setUpperRightLongitude(Double.parseDouble(tokens[3]));
+        if (tokens.length == 4) {
+            boundsSet = true;
+            this.bounds = bounds;
+            try {
+                setLowerLeftLatitude(Double.parseDouble(tokens[0].trim()));
+                setLowerLeftLongitude(Double.parseDouble(tokens[1].trim()));
+                setUpperRightLatitude(Double.parseDouble(tokens[2].trim()));
+                setUpperRightLongitude(Double.parseDouble(tokens[3].trim()));
+            } catch (NumberFormatException e) {
+                throw new ServerListParsingException(
+                        "Incorrect bound coordinates: " + e.toString());
+            }
+        } else {
+            throw new ServerListParsingException("Incorrect bound coordinates");
+        }
     }
 
     public String getLanguage() {

--- a/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/sqlite/ServersDataSource.java
+++ b/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/sqlite/ServersDataSource.java
@@ -23,7 +23,6 @@ import android.database.SQLException;
 import android.database.sqlite.SQLiteDatabase;
 import android.util.Log;
 
-import java.text.SimpleDateFormat;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -54,8 +53,6 @@ public class ServersDataSource {
             MySQLiteHelper.COLUMN_CENTER,
             MySQLiteHelper.COLUMN_ZOOM};
 
-    private SimpleDateFormat dateFormat = (SimpleDateFormat) SimpleDateFormat.getDateTimeInstance();
-
     private ServersDataSource(Context context) {
         dbHelper = MySQLiteHelper.getInstance(context);
     }
@@ -78,29 +75,35 @@ public class ServersDataSource {
     public Server createServer(Server s) {
         ContentValues values = new ContentValues();
 
-        if (s.getDate() != null) {
-            values.put(MySQLiteHelper.COLUMN_DATE, s.getDate());
-            Log.d(OTPApp.TAG,
-                    "Wrote '" + s.getRegion() + "' server date to SQLLite - " + s.getDate());
-        }
-        values.put(MySQLiteHelper.COLUMN_REGION, s.getRegion());
-        values.put(MySQLiteHelper.COLUMN_BASEURL, s.getBaseURL());
-        values.put(MySQLiteHelper.COLUMN_BOUNDS, s.getBounds());
-        values.put(MySQLiteHelper.COLUMN_CENTER, s.getCenter());
-        values.put(MySQLiteHelper.COLUMN_ZOOM, s.getZoom());
-        values.put(MySQLiteHelper.COLUMN_LANGUAGE, s.getLanguage());
-        values.put(MySQLiteHelper.COLUMN_CONTACT_NAME, s.getContactName());
-        values.put(MySQLiteHelper.COLUMN_CONTACT_EMAIL, s.getContactEmail());
+        if ((s.getRegion() != null) && (s.getBaseURL() != null) && (s.getBounds() != null)
+                && (s.getCenter() != null) && (s.getZoom() != null) && (s.getLanguage() != null)
+                && (s.getContactName() != null) && (s.getContactEmail() != null)) {
+            if (s.getDate() != null) {
+                values.put(MySQLiteHelper.COLUMN_DATE, s.getDate());
+                Log.d(OTPApp.TAG,
+                        "Wrote '" + s.getRegion() + "' server date to SQLLite - " + s.getDate());
+            }
+            values.put(MySQLiteHelper.COLUMN_REGION, s.getRegion());
+            values.put(MySQLiteHelper.COLUMN_BASEURL, s.getBaseURL());
+            values.put(MySQLiteHelper.COLUMN_BOUNDS, s.getBounds());
+            values.put(MySQLiteHelper.COLUMN_CENTER, s.getCenter());
+            values.put(MySQLiteHelper.COLUMN_ZOOM, s.getZoom());
+            values.put(MySQLiteHelper.COLUMN_LANGUAGE, s.getLanguage());
+            values.put(MySQLiteHelper.COLUMN_CONTACT_NAME, s.getContactName());
+            values.put(MySQLiteHelper.COLUMN_CONTACT_EMAIL, s.getContactEmail());
 
-        long insertId = database.insert(MySQLiteHelper.TABLE_SERVERS, null,
-                values);
-        Cursor cursor = database.query(MySQLiteHelper.TABLE_SERVERS,
-                allColumns, MySQLiteHelper.COLUMN_ID + " = " + insertId, null,
-                null, null, null);
-        cursor.moveToFirst();
-        Server newServer = cursorToServer(cursor);
-        cursor.close();
-        return newServer;
+            long insertId = database.insert(MySQLiteHelper.TABLE_SERVERS, null,
+                    values);
+            Cursor cursor = database.query(MySQLiteHelper.TABLE_SERVERS,
+                    allColumns, MySQLiteHelper.COLUMN_ID + " = " + insertId, null,
+                    null, null, null);
+            cursor.moveToFirst();
+            Server newServer = cursorToServer(cursor);
+            cursor.close();
+            return newServer;
+        } else {
+            return null;
+        }
     }
 
     public void deleteServer(Server server) {

--- a/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/tasks/ServerSelector.java
+++ b/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/tasks/ServerSelector.java
@@ -47,6 +47,7 @@ import java.util.List;
 import au.com.bytecode.opencsv.CSVReader;
 import edu.usf.cutr.opentripplanner.android.OTPApp;
 import edu.usf.cutr.opentripplanner.android.R;
+import edu.usf.cutr.opentripplanner.android.exceptions.ServerListParsingException;
 import edu.usf.cutr.opentripplanner.android.listeners.ServerCheckerCompleteListener;
 import edu.usf.cutr.opentripplanner.android.listeners.ServerSelectorCompleteListener;
 import edu.usf.cutr.opentripplanner.android.model.Server;
@@ -224,15 +225,37 @@ public class ServerSelector extends AsyncTask<LatLng, Integer, Integer>
             try {
                 Long currentTime = Calendar.getInstance().getTime().getTime();
 
-                List<String[]> entries = reader.readAll();
-                for (String[] e : entries) {
-                    if (e[0].equalsIgnoreCase("Region")) {
+                List<String[]> serversStrings = reader.readAll();
+                for (String[] serverString : serversStrings) {
+                    if (serverString[0].equalsIgnoreCase("Region")) {
                         continue; //Ignore the first line of the file
                     }
-
-                    Server s = new Server(currentTime, e[0], e[1], e[2], e[3], e[4], e[5], e[6],
-                            e[7]);
-                    serverList.add(s);
+                    for (String stringEntry : serverString) {
+                        stringEntry = stringEntry.trim();
+                    }
+                    boolean allFieldsNotNull = true;
+                    for (String serverField : serverString) {
+                        serverField = serverField.trim();
+                        if (serverField.equals("")) {
+                            Log.e(OTPApp.TAG, "Some necessary fields are null, server not added");
+                            allFieldsNotNull = false;
+                            break;
+                        }
+                    }
+                    if (allFieldsNotNull && (serverString.length >= 8)) {
+                        try {
+                            Server s = new Server(currentTime, serverString[0], serverString[1],
+                                    serverString[2], serverString[3], serverString[4],
+                                    serverString[5], serverString[6], serverString[7]);
+                            serverList.add(s);
+                        } catch (ServerListParsingException e) {
+                            Log.e(OTPApp.TAG,
+                                    "Error parsing necessary fields, server not added: " + e);
+                        }
+                    } else {
+                        Log.e(OTPApp.TAG,
+                                "Server does not provide necessary fields, server not added");
+                    }
                 }
             } catch (IOException e) {
                 Log.e(OTPApp.TAG, "Problem reading CSV server file: " + e.getMessage());
@@ -268,7 +291,9 @@ public class ServerSelector extends AsyncTask<LatLng, Integer, Integer>
     private void insertServerListToDatabase(List<Server> servers) {
         dataSource.open();
         for (Server server : servers) {
-            dataSource.createServer(server);
+            if (dataSource.createServer(server) == null) {
+                Log.e(OTPApp.TAG, "Some server fields are incorrect, server not added");
+            }
         }
         dataSource.close();
     }

